src/arch/arm64/vm.c

```diff
@@ -0,0 +1,459 @@
+    bus_init(&v->io_bus);
+    bus_init(&v->mmio_bus);
+    dev_init(&priv->iodev, ARM_IOPORT_BASE, ARM_IOPORT_SIZE, v, pio_handler);
+    bus_register_dev(&v->mmio_bus, &priv->iodev);
+
+    /* Initialize PCI bus */
+    pci_init(&v->pci);
+    v->pci.pci_mmio_dev.base = ARM_PCI_CFG_BASE;
+    bus_register_dev(&v->mmio_bus, &v->pci.pci_mmio_dev);
+
+    /* Initialize serial device */
+    if (serial_init(&v->serial, &v->io_bus))
+        return throw_err("Failed to init UART device");
+
+    if (finalize_irqchip(v) < 0)
+        return -1;
+
+    return 0;
+}
+
+/* The arm64 kernel header
+ * Reference https://docs.kernel.org/arm64/booting.html
+ */
+typedef struct {
+    uint32_t code0;       /* Executable code */
+    uint32_t code1;       /* Executable code */
+    uint64_t text_offset; /* Image load offset, little endian */
+    uint64_t image_size;  /* Effective Image size, little endian */
+    uint64_t flags;       /* kernel flags, little endian */
+    uint64_t res2;        /* reserved */
+    uint64_t res3;        /* reserved */
+    uint64_t res4;        /* reserved */
+    uint32_t magic;       /* Magic number, little endian, "ARM\x64" */
+    uint32_t res5;        /* reserved (used for PE COFF offset) */
+} arm64_kernel_header_t;
+
+int vm_arch_load_image(vm_t *v, void *data, size_t datasz)
+{
+    vm_arch_priv_t *priv = (vm_arch_priv_t *) v->priv;
+
+    arm64_kernel_header_t *header = data;
+    if (header->magic != 0x644d5241U)
+        return throw_err("Invalid kernel image\n");
+
+    uint64_t offset;
+    if (header->image_size == 0)
+        offset = 0x80000;
+    else
+        offset = header->text_offset;
+
+    if (offset + datasz >= ARM_KERNEL_SIZE ||
+        offset + header->image_size >= ARM_KERNEL_SIZE) {
+        return throw_err("Image size too large\n");
+    }
+
+    void *dest = vm_guest_to_host(v, ARM_KERNEL_BASE + offset);
+    memmove(dest, data, datasz);
+    priv->entry = ARM_KERNEL_BASE + offset;
+    return 0;
+}
+
+int vm_arch_load_initrd(vm_t *v, void *data, size_t datasz)
+{
+    vm_arch_priv_t *priv = (vm_arch_priv_t *) v->priv;
+    void *dest = vm_guest_to_host(v, ARM_INITRD_BASE);
+    memmove(dest, data, datasz);
+    priv->initrdsz = datasz;
+    return 0;
+}
+
+/* MPIDR is used by fdt generation.
+   Reference:
+ * https://developer.arm.com/documentation/ddi0601/2022-03/AArch64-Registers/MPIDR-EL1--Multiprocessor-Affinity-Register?lang=en
+ */
+#define ARM_MPIDR_BITMASK 0xFF00FFFFFFUL
+#define ARM_MPIDR_REG_ID ARM64_SYS_REG(3, 0, 0, 0, 5)
+
+static int get_mpidr(vm_t *v, uint64_t *mpidr)
+{
+    struct kvm_one_reg reg;
+    reg.addr = (uint64_t) mpidr;
+    reg.id = ARM_MPIDR_REG_ID;
+
+    if (ioctl(v->vcpu_fd, KVM_GET_ONE_REG, &reg) < 0)
+        return throw_err("Failed to get MPIDR register\n");
+
+    *mpidr &= ARM_MPIDR_BITMASK;
+    return 0;
+}
+
+/* Helper function to simplify error handling */
+#define _FDT(exp)                                                         \
```

Comment:
Furthermore, you can utilize [Variadic Macros](https://gcc.gnu.org/onlinedocs/cpp/Variadic-Macros.html) to make it more elegant.

Original
```c
_FDT(fdt_create(fdt, FDT_MAX_SIZE));
```

Revised:
```c
__FDT(create, FDT_MAX_SIZE);
```

The possible definition of the above macro would be:
```c
#define __FDT(action, ...) \
    do {
        int __ret = fdt_##action(fdt, __VA_ARGS__ );
        if (__ret < 0)
            return throw_err(...);
    } while (0)
```

By the way, the definition of `__FDT` should appear right before the function definition of `generate_fdt`, and there must be a corresponding `#undef __FDT` right after `generate_fdt`.

Reply:
Alternatively, the way what you have used for `_REG` is reasonably well. You can  apply `__FDT` in the same way. For consistency, enforce `__REG` rather than `_REG`.

Reply:
I pushed a new revision with these changes.